NXDRIVE-1436: Split test files (#783)
nuxeo/nuxeo-drive: docs/dep/2019-01 Generate a unique test environment during the setup.md

# Generate a unique test environment during the setup

- Created: 2019-01-31

> - Last-Modified: YYYY-MM-DD

Reviewer: Needs a real date :)

- Author: Mickaël Schoentgen <mschoentgen@nuxeo.com>,
          Léa Klein <lklein@uxeo.com>
- Status: draft
- Related-Ticket: [NXDRIVE-1436](https://jira.nuxeo.com/browse/NXDRIVE-1436)

## Abstract

Generate a test environment that will be unique to the run time.

> Running the test suite twice should not use same users and workspaces to ensure tests reproductilibity, reliability and debugging.

Reviewer: ```suggestion
Running the test suite twice should not use same users and workspaces to ensure tests reproducibility, reliability and debugging.
```

## Rationale

As of now, tests are calling the `NuxeoDrive.SetupIntegrationTests` operation to create users and workspaces.

With the future move to Jenkins X for testing ([NXDRIVE-1435](https://jira.nuxeo.com/browse/NXDRIVE-1435)), we will encounter errors because 3 OSes will talk to the server at the same time.
It will break everything as users and workspaces are the same everywhere.

Before that move, we need to tweak tests setup to:

- create needed user for each tests;
- create the workspace based on the OS name and the test name/function/whatever (we should keep it short for OS path limits);
- enable synchronization of that workspace for users that need it;
- clean-up everything at the end of the test.

### Idea

Using the Python client, we should be able to create what we need.

> We should no more call the `NuxeoDrive.SetupIntegrationTests` operation.

Reviewer: ```suggestion
We shouldn't call the `NuxeoDrive.SetupIntegrationTests` operation anymore.
```

## Specifications

We will completely move to `pytest` instead of having a mix with `unittest` format.

> We then may be able to fully embrace pytest [fixture's factories](https://docs.pytest.org/en/latest/fixture.html#factories-as-fixtures).

Reviewer: ```suggestion
We will then be able to fully embrace pytest [fixture's factories](https://docs.pytest.org/en/latest/fixture.html#factories-as-fixtures).
```

Speaking of factories, we may have a look at [Factory Boy](https://factoryboy.readthedocs.io/en/latest/).

> We will split the `tests` folder to reflect the actual tests file it constains.

Reviewer: ```suggestion
We will split the `tests` folder to reflect the actual test files it contains.
```

    try:
        yield path
    finally:
        with suppress(OSError):
            rmtree(path)

@pytest.fixture
def manager(tempdir):
    """Manager instance with automatic clean-up."""
    with Manager(home=tempdir) as man:
        yield man

@pytest.fixture
def user_factory(server, faker):
    """User creation factory with automatic clean-up."""
    _user = None
    fake = faker()

    def _create_user(
        password: str = "Administrator",
        lastName: str = fake.last_name(),
        firstName: str = fake.first_name(),
        email: str = fake.email(),
        company: str = fake.company(),
    ):
        nonlocal _user

        username: str = firstName.lower()
        user_ = User(
            properties={
                'lastName': lastName,
                'firstName': firstName,
                'username': username,
                'email': email,
                'company': company,
                'password': password,
            },
        )
        _user = server.users.create(user_)
        _user.password = password
        return _user

    try:
        yield _create_user
    finally:
        with suppress(Exception):
            _user.delete()
```

> This may be seen as a lot of code, but it is actually only what is needed to factorize such objects across all tests.

Reviewer: ```suggestion
This may look like a lot of code, but it is a small addition compared to what we'll be able to factorized across all tests.
```
